### ISSUE TITLE
Add level argument to filter (?) 

### ISSUE THREAD

**Issue by WillAyd** [pandas-dev/pandas#23541]
Assume we have a DataFrame that looks something as follows and we want to *exclude* the column for "Total" in the second level of the column:

```python
In [2]: mi = pd.MultiIndex.from_product([['2018'], ['Q1', 'Q2', 'Q3', 'Q4', 'Tot
   ...: al']])                                                                  

In [3]: df = pd.DataFrame([[1, 2, 3, 4, 10]], columns=mi)                       

In [4]: df                                                                      
Out[4]: 
  2018               
    Q1 Q2 Q3 Q4 Total
0    1  2  3  4    10
```

There is not an easy way to do that. One option is to add the `level` argument to `DataFrame.filter`, though there are ongoing discussions around if we want to keep or deprecate that method.

**Comment by Dr-Irv** [comment]
A couple of thoughts:

1. Maybe rename `DataFrame.filter()` to `DataFrame.filter_axis()`, because that is what you are doing
2. Maybe the title of this Issue should be "Support selecting axis labels for MultiIndex via filters"
3. One could imagine having different filters that are applied to different axes at the same time, so I don't think that just adding a `level` argument is enough.  For example, if your first level was years `2000, 2001, ..., 2018` and your second level had quarters `Q1, Q2, Q3, Q4`, you might want to pick all years between 2010 and 2015, and just the first 2 quarters, so you need a filter for each level. On the other hand, you could just string the `filter()` calls together via method chaining to do that.  So maybe the idea of specifying filters for multiple levels at once is a bit too much of an enhancement.

**Comment by alimcmaster1** [comment]
Re point (1) that seems like a scary change? All users would have to change to that new method name, filter seems like a clear method name to me, axis is just something you are filtering on.

**Comment by jreback** [comment]
- we have discussed depreciating filter in the past. filter is something you do to rows (from SQL and R)

- this is getting close to duplicating what .loc does (for multi axis filtering)

**Comment by Dr-Irv** [comment]
@jreback Not sure it duplicates what `.loc` does as it seems the advantage of `DataFrame.filter()` is that you can use `regex` or `like` arguments to select the columns you want.  The proposal here is to make `filter()` support `MultiIndex`.  If the idea is to deprecate `filter()`, then the rename I suggested makes the functionality more clear and separates it from the SQL/R interpretation.

I don't feel strongly about this as this discussion made me aware of `filter()` in the first place, and now I will start using it to select specific columns.

**Comment by jreback** [comment]
- .loc should do inexact matching (maybe by a like keyword), but it’s already pretty overloaded and you would need to differentiate between exact and inexact matches per axis - .filter already uses .loc internally FYI

- we depreciated .select a couple versions back to possibly use an alternative here - as it’s rhe correct verb for row filtering 

the confusing issue here is that you are filtering in the labels and not the data (which is what filter implies), and also note that groupby.filter does something different as well

**Comment by Dr-Irv** [comment]
@jreback I think what @WillAyd was looking for was a way to do inexact matching on the labels for columns that are backed by a `MultiIndex`.  The current `filter()` only works for columns backed by a regular `Index`. 

So one idea is to deprecate `filter()` because of the ambiguity in the name, and then have a `filter_axis()` or `filter_labels()` or `select_labels()` (or some other name) that allows inexact matching on either the row labels or column labels, even when either is backed by a `MultiIndex`.

**Comment by WillAyd** [comment]
Yep thanks @Dr-Irv . I'm not obviously not tied to `filter` being the ultimate home here. `.loc` is certainly another option as well though the pattern matching / exclusion support might be a little tricky to stuff in that method alone

**Comment by ShreyasPatel031** [comment]
Added **level** argument to filter MultiIndexes. Throws error if level is not passed for MultiIndex. Works for all items, like and regex condition.
```
>>> df = pd.DataFrame(
...             [["bare", "on", 'test'],
...              ["bar", "owner", 'test']],
...             columns=["first", "second", "third"])

>>> df = df.set_index(['first', 'second'])

>>> items = ["bare", "on", "bar"]

>>> df.filter(items, axis=0, level=1)
first second third     
bare  on      test

```

**Comment by ShreyasPatel031** [comment]
take
